File: app/src/main/java/com/adiaz/deportesmadrid/adapters/TeamMatchesAdapter.java
package com.adiaz.deportesmadrid.adapters;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.adiaz.deportesmadrid.R;
import com.adiaz.deportesmadrid.retrofit.groupsdetails.MatchRetrofit;
import com.adiaz.deportesmadrid.utils.Constants;
import com.adiaz.deportesmadrid.utils.StateAnnotation;
import com.adiaz.deportesmadrid.utils.Utils;

import org.apache.commons.lang3.StringUtils;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class TeamMatchesAdapter extends RecyclerView.Adapter<TeamMatchesAdapter.ViewHolder> {

    public static final String TAG = TeamMatchesAdapter.class.getSimpleName();

    Context mContext;
    List<MatchRetrofit> mMatchList;
    String mTeamId;

    final private ListItemClickListener mListItemClickListener;

    public TeamMatchesAdapter(Context context, List<MatchRetrofit> matchList, String mTeamId, ListItemClickListener listItemClickListener) {
        this.mContext = context;
        this.mMatchList = matchList;
        this.mTeamId = mTeamId;
        this.mListItemClickListener = listItemClickListener;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater = LayoutInflater.from(mContext);
        View view = layoutInflater.inflate(R.layout.listitem_team_match, parent, false);
        return new TeamMatchesAdapter.ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        MatchRetrofit match = mMatchList.get(position);
        holder.tvWeek.setText(mContext.getString(R.string.calendar_week, match.getNumWeek()));
        String teamLocal = Constants.FIELD_EMPTY;
        if (match.getTeamLocal()!=null && StringUtils.isNotEmpty(match.getTeamLocal().getName())) {
            teamLocal = match.getTeamLocal().getName();
        }
        String teamVisitor = Constants.FIELD_EMPTY;
        if (match.getTeamVisitor()!=null && StringUtils.isNotEmpty(match.getTeamVisitor().getName())) {
            teamVisitor = match.getTeamVisitor().getName();
        }
        /* check if a team is resting */
        if (teamLocal.equals(Constants.FIELD_EMPTY) && !teamVisitor.equals(Constants.FIELD_EMPTY)) {
            teamLocal = mContext.getString(R.string.RESTING);
        }
        if (teamVisitor.equals(Constants.FIELD_EMPTY) && !teamLocal.equals(Constants.FIELD_EMPTY)) {
            teamVisitor = mContext.getString(R.string.RESTING);
        }
        boolean isLocal = false;
        if (match.getTeamLocal() != null && match.getTeamLocal().getName().equalsIgnoreCase(mTeamId)) {
            isLocal = true;
        }
        if (isLocal) {
            holder.tvLocalOrVisitor.setText(mContext.getString(R.string.calendar_local));
            holder.tvOpponent.setText(teamVisitor);
        } else {
            holder.tvLocalOrVisitor.setText(mContext.getString(R.string.calendar_visitor));
            holder.tvOpponent.setText(teamLocal);
        }
        String dateStr = Constants.FIELD_EMPTY;
        if (match.getDate()!=null && match.getState()!=StateAnnotation.DESCANSA) {
            DateFormat dateFormat = new SimpleDateFormat(Constants.DATE_FORMAT);
            dateStr = dateFormat.format(match.getDate());
        }
        holder.tvDate.setText(dateStr);
        String placeName = Constants.FIELD_EMPTY;
        if (match.getPlace()!=null) {
            placeName = match.getPlace().getName();
        }
        holder.tvPlace.setText(placeName);
        holder.tvState.setText(Utils.getStringResourceByName(mContext, StateAnnotation.stringKey(match.getState())));
        if (match.getState()== StateAnnotation.FINALIZADO || match.getState() == StateAnnotation.NO_PRESENTADO) {
            holder.tvScore.setText(mContext.getString(R.string.calendar_score, match.getScoreLocal(), match.getScoreVisitor()));
            holder.tvScore.setVisibility(View.VISIBLE);
            int scoreTeam;
            int scoreTeamOpponent;
            if (isLocal) {
                scoreTeam = match.getScoreLocal();
                scoreTeamOpponent = match.getScoreVisitor();
            } else {
                scoreTeam = match.getScoreVisitor();
                scoreTeamOpponent = match.getScoreLocal();
            }
            holder.ivThumbVictory.setVisibility(View.VISIBLE);
            if (scoreTeam > scoreTeamOpponent) {
                holder.ivThumbVictory.setImageResource(R.drawable.ic_thumb_up);
            } else if(scoreTeam<scoreTeamOpponent) {
                holder.ivThumbVictory.setImageResource(R.drawable.ic_thumb_down);
            } else {
                holder.ivThumbVictory.setImageResource(R.drawable.ic_thumbs_up_down);
            }
        } else {
            holder.tvScore.setVisibility(View.INVISIBLE);
            holder.ivThumbVictory.setVisibility(View.INVISIBLE);
        }
    }

    @Override
    public int getItemCount() {
        return mMatchList.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        @BindView(R.id.cl_teamcalendar)
        View vTeamCalendar;

        @BindView(R.id.tv_week)
        TextView tvWeek;

        @BindView(R.id.tv_local_visitor)
        TextView tvLocalOrVisitor;

        @BindView(R.id.tv_opponent)
        TextView tvOpponent;

        @BindView(R.id.tv_date)
        TextView tvDate;

        @BindView(R.id.tv_place)
        TextView tvPlace;

        @BindView(R.id.tv_state)
        TextView tvState;

        @BindView(R.id.tv_score)
        TextView tvScore;

        @BindView(R.id.iv_thumb_victory)
        ImageView ivThumbVictory;

        public ViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
            vTeamCalendar.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            Log.d(TAG, "onClick: ");
            mListItemClickListener.onListItemClick(getAdapterPosition());
        }
    }

    public interface ListItemClickListener {
        void onListItemClick(int clickedItemIndex);
    }
}

File: app/src/main/java/com/adiaz/deportesmadrid/adapters/TeamMatchesAdapter.kt
package com.adiaz.deportesmadrid.adapters

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.adiaz.deportesmadrid.R
import com.adiaz.deportesmadrid.extensions.layoutInflater
import com.adiaz.deportesmadrid.retrofit.groupsdetails.MatchRetrofit
import com.adiaz.deportesmadrid.utils.Constants
import com.adiaz.deportesmadrid.utils.StateAnnotation
import com.adiaz.deportesmadrid.utils.Utils
import kotlinx.android.synthetic.main.listitem_team_match.view.*
import org.apache.commons.lang3.StringUtils
import java.text.SimpleDateFormat

class TeamMatchesAdapter(
        private var mContext: Context,
        private var mMatchList: List<MatchRetrofit>,
        private var mTeamId: String,
        private val mListItemClickListener: ListItemClickListener) : RecyclerView.Adapter<TeamMatchesAdapter.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TeamMatchesAdapter.ViewHolder {
        val view = mContext.layoutInflater.inflate(R.layout.listitem_team_match, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val match = mMatchList[position]
        holder.tvWeek.text = mContext.getString(R.string.calendar_week, match.numWeek)
        var teamLocal = Constants.FIELD_EMPTY
        if (match.teamLocal != null && StringUtils.isNotEmpty(match.teamLocal.name)) {
            teamLocal = match.teamLocal.name
        }
        var teamVisitor = Constants.FIELD_EMPTY
        if (match.teamVisitor != null && StringUtils.isNotEmpty(match.teamVisitor.name)) {
            teamVisitor = match.teamVisitor.name
        }
        /* check if a team is resting */
        if (teamLocal == Constants.FIELD_EMPTY && teamVisitor != Constants.FIELD_EMPTY) {
            teamLocal = mContext.getString(R.string.RESTING)
        }
        if (teamVisitor == Constants.FIELD_EMPTY && teamLocal != Constants.FIELD_EMPTY) {
            teamVisitor = mContext.getString(R.string.RESTING)
        }
        var isLocal = false
        if (match.teamLocal != null && match.teamLocal.name.equals(mTeamId, ignoreCase = true)) {
            isLocal = true
        }
        if (isLocal) {
            holder.tvLocalOrVisitor.text = mContext.getString(R.string.calendar_local)
            holder.tvOpponent.text = teamVisitor
        } else {
            holder.tvLocalOrVisitor.text = mContext.getString(R.string.calendar_visitor)
            holder.tvOpponent.text = teamLocal
        }
        var dateStr = Constants.FIELD_EMPTY
        if (match.date != null && match.state != StateAnnotation.DESCANSA) {
            val dateFormat = SimpleDateFormat(Constants.DATE_FORMAT)
            dateStr = dateFormat.format(match.date)
        }
        holder.tvDate.text = dateStr
        var placeName = Constants.FIELD_EMPTY
        if (match.place != null) {
            placeName = match.place.name
        }
        holder.tvPlace.text = placeName
        holder.tvState.text = Utils.getStringResourceByName(mContext, StateAnnotation.stringKey(match.state!!))
        if (match.state == StateAnnotation.FINALIZADO || match.state == StateAnnotation.NO_PRESENTADO) {
            holder.tvScore.text = mContext.getString(R.string.calendar_score, match.scoreLocal, match.scoreVisitor)
            holder.tvScore.visibility = View.VISIBLE
            val scoreTeam: Int
            val scoreTeamOpponent: Int
            if (isLocal) {
                scoreTeam = match.scoreLocal!!
                scoreTeamOpponent = match.scoreVisitor!!
            } else {
                scoreTeam = match.scoreVisitor!!
                scoreTeamOpponent = match.scoreLocal!!
            }
            holder.ivThumbVictory.visibility = View.VISIBLE
            if (scoreTeam > scoreTeamOpponent) {
                holder.ivThumbVictory.setImageResource(R.drawable.ic_thumb_up)
            } else if (scoreTeam < scoreTeamOpponent) {
                holder.ivThumbVictory.setImageResource(R.drawable.ic_thumb_down)
            } else {
                holder.ivThumbVictory.setImageResource(R.drawable.ic_thumbs_up_down)
            }
        } else {
            holder.tvScore.visibility = View.INVISIBLE
            holder.ivThumbVictory.visibility = View.INVISIBLE
        }
    }

    override fun getItemCount(): Int {
        return mMatchList.size
    }

    inner class ViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        var vTeamCalendar: View = itemView.cardView
        var tvWeek: TextView = itemView.tvWeek
        var tvLocalOrVisitor: TextView = itemView.tvLocalVisitor
        var tvOpponent: TextView = itemView.tvOpponent
        var tvDate: TextView = itemView.tvDate
        var tvPlace: TextView = itemView.tvPlace
        var tvState: TextView = itemView.tvState
        var tvScore: TextView = itemView.tvScore
        var ivThumbVictory: ImageView = itemView.ivThumbVictory

        init {
            vTeamCalendar.setOnClickListener(this)
        }

        override fun onClick(view: View) {
            mListItemClickListener.onListItemClick(adapterPosition)
        }
    }

    interface ListItemClickListener {
        fun onListItemClick(clickedItemIndex: Int)
    }

    companion object {

        val TAG = TeamMatchesAdapter::class.java.simpleName
    }
}
